lib/serializer/json/aws_iot_serializer_json_encoder.c

```diff
@@ -0,0 +1,691 @@
+        .getExtraBufferSizeNeeded = _getExtraBufferSizeNeeded,
+        .init = _init,
+        .destroy = _destroy,
+        .openContainer = _openContainer,
+        .openContainerWithKey = _openContainerWithKey,
+        .closeContainer = _closeContainer,
+        .append = _append,
+        .appendKeyValue = _appendKeyValue
+};
+
+static void _stopContainer( _jsonContainer_t* pContainer, AwsIotSerializerDataType_t containerType )
+{
+    if( containerType == AWS_IOT_SERIALIZER_CONTAINER_ARRAY )
+    {
+        pContainer->pBuffer[ pContainer->offset++ ] = _JSON_ARRAY_END_CHAR;
+    }
+    else if( containerType == AWS_IOT_SERIALIZER_CONTAINER_MAP )
+    {
+        pContainer->pBuffer[ pContainer->offset++ ] = _JSON_OBJECT_END_CHAR;
+    }
+}
+static size_t _getSerializedLength(
+        _jsonContainer_t* pContainer,
+        AwsIotSerializerDataType_t dataType,
+        AwsIotSerializerScalarData_t* pScalarData )
+
+{
+    switch( dataType )
+    {
+        case AWS_IOT_SERIALIZER_CONTAINER_MAP:
+        case AWS_IOT_SERIALIZER_CONTAINER_ARRAY:
+            return _jsonEmptyContainerLength;
+        case AWS_IOT_SERIALIZER_CONTAINER_STREAM:
+            break;
+        case AWS_IOT_SERIALIZER_SCALAR_TEXT_STRING:
+            return _jsonStringLength( pScalarData->value.stringLength );
+        case AWS_IOT_SERIALIZER_SCALAR_BYTE_STRING:
+            return _jsonByteStringLength( pScalarData->value.stringLength );
+        case AWS_IOT_SERIALIZER_SCALAR_SIGNED_INT:
+            return _jsonIntegerLength( pScalarData->value.signedInt );
+        case AWS_IOT_SERIALIZER_SCALAR_BOOL:
+            return _jsonBoolLength( pScalarData->value.booleanValue );
+        case AWS_IOT_SERIALIZER_SCALAR_NULL:
+            return _JSON_NULL_VALUE_LENGTH;
+        default:
+            break;
+    }
+
+    return 0;
+}
+
+static size_t _getKeyValueLength(
+        _jsonContainer_t* pContainer,
+        size_t keyLength,
+        AwsIotSerializerDataType_t valueType,
+        AwsIotSerializerScalarData_t* pScalarValue )
+{
+    size_t keyValueLength = 0, valueLength;
+
+    valueLength =  _getSerializedLength( pContainer, valueType, pScalarValue );
+    keyValueLength = _jsonKeyValuePairLength( keyLength, valueLength );
+
+    if( !pContainer->isEmpty )
+    {
+        keyValueLength += 1;
+    }
+
+    return keyValueLength;
+}
+
+static size_t _getValueLength(
+        _jsonContainer_t* pContainer,
+        AwsIotSerializerDataType_t valueType,
+        AwsIotSerializerScalarData_t* pScalarValue )
+{
+    size_t length = _getSerializedLength( pContainer, valueType, pScalarValue );
+    if( !pContainer->isEmpty )
+    {
+        length += 1;
+    }
+    return length;
+}
+
+static void _appendTextString( _jsonContainer_t* pContainer, const char* pStr, size_t strLength )
+{
+    pContainer->pBuffer[ pContainer->offset++ ] = _JSON_STRING_WRAPPER;
+    memcpy( ( pContainer->pBuffer +  pContainer->offset ), pStr, strLength );
```

Comment:
just wonder why not strcpy

Reply:
strcpy has the risk of buffer overrun if pStr is not null terminated. Also it strcpy/strncpy will include an implicit null termination character in success case. We want to avoid that as strings in json are quotes delimited.